File: frameworks/dotnet/aspnet-mvc/UnitTests/ErrorsTests.cs
namespace UnitTests;

/// <summary>
/// errors: every refusal is ASP.NET Core's own, written as ProblemDetails. Routing's 404 and 405
/// get theirs from AddProblemDetails, and [ApiController] writes its own for the action's
/// NotFound and for a body it cannot read.
/// </summary>
public sealed class ErrorsTests(MvcApp app) : IClassFixture<MvcApp>
{
    // rb:test errors.unmatched
    [Fact]
    [Trait("corpus", "errors.unmatched")]
    public async Task A_path_no_route_matches_is_the_routers_404()
    {
        using HttpResponseMessage response = await app.CreateClient().GetAsync("/errors/unmatched");

        await Problem(HttpStatusCode.NotFound, response);
    }

    // rb:test errors.not_found
    [Fact]
    [Trait("corpus", "errors.not_found")]
    public async Task An_id_with_no_row_is_the_actions_404()
    {
        using HttpResponseMessage response = await app.CreateClient().GetAsync("/items/999999");

        await Problem(HttpStatusCode.NotFound, response);
    }

    // rb:test errors.wrong_method
    [Fact]
    [Trait("corpus", "errors.wrong_method")]
    public async Task A_method_the_path_has_no_action_for_is_405()
    {
        using HttpResponseMessage response = await app.CreateClient().PostAsync("/items/17", null);

        await Problem(HttpStatusCode.MethodNotAllowed, response);
        Assert.Equal("DELETE, GET, HEAD, PATCH, PUT", Answer.Header(response, "allow"));
    }

    // rb:test errors.malformed
    [Fact]
    [Trait("corpus", "errors.malformed")]
    public async Task A_body_that_is_not_json_is_the_automatic_400()
    {
        using HttpResponseMessage response = await app.CreateClient().PostAsync("/body/validate/small", new StringContent("""{"customerId": 1, "lines": [""", null, "application/json"));

        JsonNode problem = await Problem(HttpStatusCode.BadRequest, response);
        Assert.Equal(["$.lines[0]", "order"], problem["errors"]!.AsObject().Select(e => e.Key).Order(StringComparer.Ordinal));
    }

    private static async Task<JsonNode> Problem(HttpStatusCode status, HttpResponseMessage response)
    {
        Assert.Equal(status, response.StatusCode);
        Assert.Equal("application/problem+json", response.Content.Headers.ContentType?.MediaType);
        JsonNode problem = await Answer.Json(response);
        Assert.Equal((int)status, (int?)problem["status"]);
        return problem;
    }
}

File: frameworks/java/spring-boot/UnitTests/src/test/java/unittests/ErrorsTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Set;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

/** errors: every refusal is Spring's own, and Boot's error page writes each body. */
class ErrorsTests extends SpringApp {

    // rb:test errors.unmatched
    @Test
    @Tag("errors.unmatched")
    void aPathNoHandlerMatchesIsBootsErrorPage404() throws Exception {
        errorPage(404, get("/errors/unmatched"));
    }

    // rb:test errors.not_found
    @Test
    @Tag("errors.not_found")
    void anIdWithNoRowIsTheHandlers404() throws Exception {
        errorPage(404, get("/items/999999"));
    }

    // rb:test errors.wrong_method
    @Test
    @Tag("errors.wrong_method")
    void aMethodThePathHasNoHandlerForIs405() throws Exception {
        HttpResponse<byte[]> response = send("POST", "/items/17", "application/json", new byte[0]);

        errorPage(405, response);
        assertEquals(Set.of("GET", "PUT", "PATCH", "DELETE"), Set.of(Answer.header(response, "allow").split(",\\s*")));
    }

    // rb:test errors.malformed
    @Test
    @Tag("errors.malformed")
    void aBodyThatIsNotJsonIsJacksons400() throws Exception {
        byte[] body = "{\"customerId\": 1, \"lines\": [".getBytes(StandardCharsets.UTF_8);

        errorPage(400, send("POST", "/body/validate/small", "application/json", body));
    }

    private static void errorPage(int status, HttpResponse<byte[]> response) {
        assertEquals(status, response.statusCode());
        assertEquals("application/json", Answer.header(response, "content-type"));
        assertEquals(status, Answer.json(response).get("status").asInt());
    }
}
